Allow adding & removing node neighbors directly

christarazi opened the pull request:
See commit msgs.

TODO: 
  - [x] Implement removing node neigh
  - [x] Test removing node neigh from the CLI
  - [x] Ensure that all the API errors for inserting and removing make sense (I simply copied them from another API endpoint)
  - [x] Consider whether we should refactor `insertNeighbor()` / `deleteNeighbor()` to return errors so that we can surface it back to the user in the API. Currently, no error is shown to the user if an insert or remove fails; only an agent log msg.

Fixes https://github.com/cilium/cilium/issues/14203

christarazi committed:
api: Add ability to insert & remove node neighbor

This functionality is useful when running in LB-only mode
(`--datapath-mode=lb-only`) because it doesn't run with K8s. This means
that there is no mechanism to discover nodes.

Therefore, this commit allows the operator to add or remove nodes from a
load-balancing cluster.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
cmd: Add cilium node neigh CLI

This allows users to manage the current node's neighbor table (the node
that's receiving this request, i.e. the node running Cilium) from the
CLI. The only supported operations at the moment are inserting and
removing a node into / from the current node's neighbor table. See
previous commit for context into why we are exposing this API.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
datapath/linux: Extract iface name detection logic

This is useful for upcoming commits where this logic would have been
duplicated.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
datapath: Extend node handler interface for neigh

This commits extends the node handler interface to handle node neighbor
operations, such as adding or removing node neighbor entry into / from
the current node neighbor table.

The ability to add / remove node neighbors will follow in subsequent
commits.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

brb reviewed (commented):
Looks great!

> Consider whether we should refactor insertNeighbor() / deleteNeighbor() to return errors

Yeah, let's make it to return an error.

christarazi committed:
datapath/linux: Refactor {insert,delete}Neighbor()

This commit refactors inserting and deleting neighbors to return errors.
This will become useful in future commits to expose the ability to add /
remove node neighbors via an API because the errors will be returned to
the user.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
datapath/linux: Add debug log for neigh operations

This is useful to validate whether a node neighbor was actually inserted
or removed.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -811,6 +805,14 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 	return nil
 }
 
+func (n *linuxNodeHandler) getApproriateIfaceName() string {
```
Typo in the function name? Also I would suggest to have a better less vague name for this function.

christarazi replied:
Thanks, fixed.

aanm commented on cilium/cmd/node_neigh_insert.go:
```diff
@@ -0,0 +1,56 @@
+var nodeNeighInsertCmd = &cobra.Command{
+	Use:     "insert <neigh name> <neigh IP>",
+	Short:   "Insert node as a neighbor into current node's neighbor table",
+	Example: `cilium node neigh insert "node1" 10.10.10.10`,
```
s/`10.10.10.10`/`192.0.2.1`/

aanm commented on cilium/cmd/node_neigh_remove.go:
```diff
@@ -0,0 +1,58 @@
+	Use:     "remove <neigh name> <neigh IP>",
+	Aliases: []string{"rm"},
+	Short:   "Remove node as a neighbor from current node's neighbor table",
+	Example: `cilium node neigh remove "node1" 10.10.10.10`,
```
s/`10.10.10.10`/`192.0.2.1`/

christarazi commented on daemon/cmd/node.go:
```diff
@@ -0,0 +1,106 @@
+	newNode := nodeTypes.Node{
+		IPAddresses: []nodeTypes.Address{
+			{
+				Type: addressing.NodeInternalIP, // TODO(christarazi): Is this correct?
```
Just pointing this out for reviewers. Would like to confirm that this makes sense. I assume that all node neighbors' IP will be an internal IP, but I'm also unsure if it matters at all. In the case that it doesn't matter, then leaving it as internal is fine and actually preferable because grabbing the node IP via `node.GetNodeIP()` (used by insert / delete neighbor code) considers the internal IP first.

brb replied:
Yeah, `addressing.NodeInternalIP` makes sense.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -620,21 +620,31 @@ func (n *linuxNodeHandler) insertNeighbor(newNode *nodeTypes.Node, ifaceName str
 // Must be called with linuxNodeHandler.mutex held.
-func (n *linuxNodeHandler) deleteNeighbor(oldNode *nodeTypes.Node) {
+func (n *linuxNodeHandler) deleteNeighbor(oldNode *nodeTypes.Node) error {
 	nextHopStr, found := n.neighNextHopByNode[oldNode.Identity()]
```
@brb In an offline sync, you mentioned that we should allow the user to specify the node identity. Initially I assumed this meant the numeric identity, but the insert / delete neighbor code doesn't use the numeric identity. Instead it uses the `(*Node).Identity()` which is a combo of the node name and the cluster ID. 

Given this, I've made the API accept a node name. Is this what we want?

brb replied:
For now it's OK to make the identity from the node name. We can always extend it if needed (by adding cluster name as a param).

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -568,7 +568,7 @@ func (n *linuxNodeHandler) insertNeighbor(newNode *nodeTypes.Node, ifaceName str
 	routes, err := netlink.RouteGet(nextHopIPv4)
 	if err != nil {
 		scopedLog.WithError(err).Error("Failed to retrieve route for remote node IP")
```
Just a small note that I forgot in my previous review. Since we are now returning errors, wouldn't it make sense to log the errors in the caller instead of potentially duplicating the log messages? i.e., one would be printed in this line and the other would again printed in the caller.

christarazi replied:
Good point, will fix.

brb commented on pkg/datapath/linux/node.go:
```diff
@@ -811,6 +805,16 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 	return nil
 }
 
+// getExternalFacingDeviceName returns the external / cluster-facing device
```
Hmm, not sure about naming of this method. In the case of `DirectRoutingDevice` it might be not external facing device, as it's used for internal communication between k8s nodes.

christarazi replied:
Any ideas for a name? I guess what I don't know is why we also include the `EncryptInterface` as well. Identifying whatever the `DirectRoutingDevice` and `EncryptInterface` have in common would help us in finding a name.

brb replied:
Maybe just `getNodeDeviceName`?

brb commented on daemon/cmd/node.go:
```diff
@@ -0,0 +1,106 @@
+			Type: addressing.NodeInternalIP,
+			IP:   ip4,
+		})
+	} else if ip6 := ip.To16(); ip6 != nil {
```
At the moment, we don't need to handle IPv6, as arping is used only for IPv4, and equivalent-based one (NPD) is currently not implemented. Maybe just return an error?

brb reviewed (commented):
Thanks, some minor comments.
